File: python/max-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        N = len(nums)
        opt = [0] * N
        opt[0] = nums[0]
        for i in range(1, N):
            opt[i] = max(nums[i], nums[i] + opt[i-1])
            
        return max(opt)

File: java/max-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int n = nums.length;
        int currProd = 1;
        int result = -10;
        for(int i=0; i<n; i++) {
            currProd = currProd * nums[i];
            if(currProd > result) {
                result = currProd;
            }
            if(currProd == 0) {
                currProd = 1;
            }
        }
        currProd = 1;

        for(int i=n-1; i>=0; i--) {
            currProd = currProd * nums[i];
            if(currProd > result) {
                result = currProd;
            }
            if(currProd == 0) {
                currProd = 1;
            }
        }
        return result;
    }
}
